### PR TITLE
feat(controller): Use different container runtime executors for each workflow. Close #4254

### LINKED ISSUE #4254: Per workflow container runtime executor setting
# Summary

Allow users to change this at runtime.

# Use Cases

To test out other executors.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by alexec**
Adds `containerRuntimeExecutors` to `workflow-controller-configmap` so you can select the executor you want to use.

This is controlled by the operation - users cannot change this setting.

The operator can allow 1 or more executors. 

Typically I'd expect this to be used for:

* Trying out `k8sapi` or `emissary` executors.
* Trying out any executor prior to switching the default.
* Running some workflows more or less securely.

```yaml
  containerRuntimeExecutor: docker
  containerRuntimeExecutors: |
    - name: k8sapi
      selector:
        matchLabels:
          workflows.argoproj.io/container-runtime-executor: k8sapi
```

### CODE REVIEW

**Review by terrytangyuan — COMMENTED** [review]
This is great to have. We were thinking about something similar along the line as well but distributing based on percentage of the total number of workflows.

File: config/config.go [hunk 1]
```diff
@@ -99,6 +101,14 @@ type Config struct {
 	InitialDelay metav1.Duration `json:"initialDelay,omitempty"`
 }
 
+func (c Config) GetContainerRuntimeExecutor(labels labels.Labels) (string, error) {
+	name, err := c.ContainerRuntimeExecutors.Select(labels)
+	if name != "" {
+		return name, err
```
Comment (terrytangyuan) [inline config/config.go:107]: Looks like err is always nil in this case so return `name, nil` would be more readable?

Comment (alexec) [reply config/config.go:107]: I know it is odd looking, but I think it is correct.

Comment (alexec) [reply config/config.go:107]: this is because it can return `"", nil` - i.e. it did not find a CRE, but there was not an error

**Comment by alexec** [comment]
@sarabala1979 can I request your review as I think @jessesuen  is busy

**Comment by alexec** [comment]
@sarabala1979 I accidentally closed this. Can you please review/approve?

File: config/config.go [hunk 2]
```diff
@@ -99,6 +101,14 @@ type Config struct {
 	InitialDelay metav1.Duration `json:"initialDelay,omitempty"`
 }
 
+func (c Config) GetContainerRuntimeExecutor(labels labels.Labels) (string, error) {
+	name, err := c.ContainerRuntimeExecutors.Select(labels)
+	if name != "" {
```
Comment (sarabala1979) [inline config/config.go:106]: can we log the error if it is not nil?

File: config/container_runtime_executors.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+		}
+		if ok {
+			return c.Name, nil
+		}
+	}
+	return "", nil
+}
+
+type ContainerRuntimeExecutor struct {
+	Name     string               `json:"name"`
```
Comment (sarabala1979) [inline config/container_runtime_executors.go:24]: the `Name` is kind of confusing.  It should be `executor` or` executorName` or `executorType`

**Comment by sarabala1979** [comment]
I think in the future we can have `defaultExecuter` under ContainerRuntimeExecutors and deprecate `ContainerRuntimeExecutor`

File: config/config.go (hunk 2 above)
Comment (alexec) [reply config/config.go:106]: I'm going to re-write this confusing code to make it clearer and less "clever"

File: config/container_runtime_executors.go (hunk 3 above)
Comment (alexec) [reply config/container_runtime_executors.go:24]: I think that "type" and "name" are synonymous. I.e. there is only one executor of type "pns", it is the "pns" executor.

I'm not sure about the correct term here, but I'd like to leave this called "name" unless you feel strongly?

**Comment by alexec** [comment]
> I think in the future we can have `defaultExecuter` under ContainerRuntimeExecutors and deprecate `ContainerRuntimeExecutor`

I don't think I should make this change. I'd have to write additional code and deal with what happens when they're both defined, but this does not add any new capabilities but does force migration on our users at some point.

**Review by sarabala1979 — APPROVED** [review]
Above comments minor

File: examples/selected-executor-workflow.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
+apiVersion: argoproj.io/v1alpha1
```
Comment (alexec) [inline examples/selected-executor-workflow.yaml:1]: @sarabala1979 @simster7 this is a fully-filled out example with a test, description, and version.

Comment (alexec) [reply examples/selected-executor-workflow.yaml:1]: @jessesuen @terrytangyuan FYI

**Pull request merged**